Newletter API 구현

anaconda77 opened the pull request:
# 개요

- Newsletter 도메인을 확장하고, Newsletter API를 구현하였음

# 배경

- TS 서버에 구현된 Newsletter 관련 API들을 스프링 서버로 이전 작업이 필요하였음

# 변경된 점
- Newsletter API 구현
  - Newsletter 리스트 조회
  - Newsletter 상세정보 조회
  - ~~Newsletter 추천 리스트 조회~~
    - UserCategory 도메인 정의가 필요, 현재 이슈에서 처리하기엔 범위가 너무 커진다고 판단하여 다음 이슈로 넘겨 처리할 예정
- Newsletter 도메인 확장
  - erd에 명세된 대로 Newsletter 도메인을 정의
- ~~Newsletter 도메인 Entity 삭제~~
  - ~~Newsletter 도메인 Entity에 아무런 비즈니스 로직이 없는 상황,  도메인 Entity의 존재 가치에 대한 의문이 듦~~
  - ~~도메인 객체는 행동하는 주체가 되어야 의미가 있다고 생각하는데, 현재는 단지 값을 저장하고 있는 VO에 불과함~~
  - ~~따라서 도메인 Entity 및 EntityDTO를 삭제하고, Repository<->Service 계층 간 JPA Entity를 직접 전달하도록 로직을 변경함~~
- Newsletter 테스트 코드 작성
  - NewsletterFixture 클래스를 정의하여, 여러 테스트 클래스에서 데모 데이터를 편리하게 생성할 수 있도록 함
- UUID Converter 유틸 클래스 정의
  - String에서 UUID으로 id 값의 타입 변환이 필요한 곳에서 사용할 수 있도록 common 패키지에 구현
  - UUID 변환 실패에 관한 예외 정의
    - IllegalArgumentException을 던지는 형태, UUID 변환 오류임을 명세하여 응답하도록 핸들러 메서드 추가
      <img width="762" alt="스크린샷 2025-04-22 오후 10 27 10" src="https://github.com/user-attachments/assets/d639950f-a473-45a1-b701-bebc51b44713" />
- 테스트 리팩터링
  - Newsletter 도메인의 변경으로 컴파일 오류가 발생하는 기존 테스트 코드들을 리팩터링
## 참고자료


## 관련 이슈

close #49

anaconda77 committed:
Feat: Newsletter 도메인 확장 및 DTO 재정의

anaconda77 committed:
Merge branch 'refs/heads/dev' into Feat/issue-#49

anaconda77 committed:
Feat: Newsletter Color Enum 정의

anaconda77 committed:
Refactor: NewsletterEnitty에 Color Enum 적용

anaconda77 committed:
Refactor: NewsletterEntity 생성자 재정의

anaconda77 committed:
Feat: Newsletter Repository 계층 정의

anaconda77 committed:
Fix: 잘못된 id 타입 수정 및 기존 Subscription 구현 부분 주석 처리(나중에 구현할 부분, 현재 컴파일 오류가 발생하므로 일단 주석 처리해둠)

anaconda77 committed:
Feat: Newsletter 도메인 엔티티, dto, jpa 엔티티 생성자 추가

anaconda77 committed:
Revert: 잘못 주석처리한 부분 되돌림

anaconda77 committed:
Test: Newsletter Fixture 정의

anaconda77 committed:
Test: 기존 Newsletter 관련 테스트 코드 옳게 동작하도록 리팩토링

anaconda77 committed:
Revert: 본래 생성자 변경 및 기존 생성자 Builder 패턴으로 변경

anaconda77 committed:
Revert: Newsletter ~ Category 연관관계 복구

anaconda77 committed:
Feat: Newsletter 도메인 Category 의존, NewsletterEntityDto CategoryEntityDto 의존하도록 변경

anaconda77 committed:
Refactor: Newsletter 도메인/DTO 삭제

anaconda77 committed:
Test: 테스트 리팩토링

anaconda77 committed:
Merge branch 'refs/heads/dev' into Feat/issue-#49

anaconda77 committed:
Feat: 뉴스레터 리스트 조회 API 구현

anaconda77 committed:
Feat: 뉴스레터 상세정보 조회 API 구현

anaconda77 committed:
Feat: UUID string->uuid 타입 변환 실패 관련 핸들러 메서드 정의

anaconda77 committed:
Test: NewsletterRepoImpl 테스트 코드 작성

anaconda77 committed:
Test: NewsletterRepoImpl 테스트 코드 작성

anaconda77 committed:
Test: NewsletterService 테스트 코드 작성

anaconda77 committed:
Feat: Newsletter API 스웨거 적용

anaconda77 committed:
Fix: UUID null인 경우도 통과하도록 수정

anaconda77 committed:
Refactor: UUID 변환 로직 Service로 이전

anaconda77 committed:
Feat: UUID 변환 유틸 클래스 분리

anaconda77 committed:
Test: NewsletterService 테스트 추가

GitJIHO commented on src/main/java/com/newzet/api/newsletter/controller/NewsletterController.java:
```diff
@@ -0,0 +1,54 @@
+		description = "뉴스레터 이름 혹은 카테고리 id로 뉴스테러 리스트를 조회한다.")
+	public ResponseEntity<SuccessResponse<NewsletterListResponse>> getNewsletterListByNameOrCategoryId(
+		@RequestParam(value = "name", required = false) String name,
+		@RequestParam(value = "categoryId", required = false) String categoryId) {
+		NewsletterListResponse newsletterListResponse = newsletterService.searchNewsletterListByNameOrCategoryId(
+			name, categoryId);
+		SuccessResponse<NewsletterListResponse> response = SuccessResponse.create(
+			ResponseCode.SUCCESS, "뉴스레터 목록 조회 성공", newsletterListResponse);
+
+		return ResponseEntity.ok(response);
+	}
+
+	@GetMapping("/{newsletterId}")
+	@Operation(summary = "뉴스레터 상세정보 조회",
+		description = "뉴스레터 id로 뉴스테러를 조회한다.")
+	public ResponseEntity<SuccessResponse<NewsletterInfoResponse>> getNewsletterById(
+		@PathVariable("newsletterId") String newsletterId) {
```
catetoryId를 String타입이 아닌, UUID 형식으로 프론트로부터 받는다면 따로 백엔드에서 변환처리를 안해도 된다는 장점이 있을 것 같은데, 아무래도 API명세에 있는 형식을 바꾸는건 지양하기 위해 해당 방식을 선택한 것이라 이해하면 될까요?

GitJIHO commented on src/main/java/com/newzet/api/common/util/UuidConverter.java:
```diff
@@ -0,0 +1,18 @@
+
+import com.newzet.api.common.util.exception.UuidConvertFailException;
+
+public class UuidConverter {
+	public static UUID convert(String stringId) {
+		if (stringId != null && !stringId.isEmpty()) {
+			try {
+				return UUID.fromString(stringId);
+			} catch (IllegalArgumentException e) {
+				throw new UuidConvertFailException(e.getMessage());
+			}
+		}
+		return null;
+	}
+}
```
좋습니다. auth 로직에서도 사용할 수 있을 것 같네요 👍

GitJIHO reviewed (commented):
고생하셨습니다 👍 

> Newsletter 도메인 Entity 삭제에 관해

현재 Newsletter 도메인 객체가 별다른 비즈니스 로직 없이 단순한 값 객체로만 사용되고 있어, 해당 객체를 삭제하고 JPA Entity를 직접 사용하는 방식으로 정리하신 부분은 충분히 이해됩니다.

다만 이렇게 되면 repository 패키지 내의 entity가 도메인 역할까지 담당하게 되어, 향후 Newsletter에 비즈니스 로직이 추가될 경우 해당 로직이 entity에 직접 들어가거나 service에서 처리되면서 구조가 다소 무거워질 수 있을 것 같다는 우려가 듭니다.

또한 다른 도메인들에서는 도메인 객체를 별도로 유지하고 있는 것으로 알고 있는데, Newsletter만 다른 방식으로 가져가게 되면 구조적인 일관성이 깨지지 않을까 하는 약간의 걱정이 있네요..!

장기적으로는 도메인 객체를 유지하는 방향도 고려해볼 수 있을 것 같습니다.

anaconda77 committed:
Revert: Newsletter/Newsletter 도메인 복구 및 EntityDTO를 통한 통신 구조 도입

anaconda77 committed:
Test: 테스트 리팩토링

anaconda77 committed:
Refactor: NewsletterEntity toEntityDto()에 null 체크 구문 추가

현재 뉴스레터 save 로직이 남아있으므로, 기존 save 후 toEntityDto()를 호출할때 category가 null이기 때문에 테스트 코드에서 예외가 발생하므로 임시 방편으로 null도 통과하도록함
추후 피드백을 받아 수정 예정

anaconda77 committed:
Refactor: toEntity() 제거 및 SubscriptionRepo save 로직 연관관계 Entity findBy로 가져오는 것으로 변경

Gyu-won commented on src/main/java/com/newzet/api/common/util/UuidConverter.java:
```diff
@@ -0,0 +1,18 @@
+				throw new UuidConvertFailException(e.getMessage());
+			}
+		}
+		return null;
```
값이 null 이거나 empty인 경우에도 UuidConvertFailException을 던지는건 어떨까요?

anaconda77 replied:
searchNewsletterListByNameOrCategoryId 의 경우에는 Name과 categoryId 검색 조건을 한 메서드로 같이 쓰는 구조인데, categoryId가 null인 상황을 고려하여 null 값 입력은 그대로 통과하도록 하였습니다!

Gyu-won commented on src/main/java/com/newzet/api/newsletter/business/NewsletterService.java:
```diff
@@ -34,18 +36,40 @@ public class NewsletterService {
 		return cacheUtil.get(CACHE_DOMAIN_PREFIX + domain, NewsletterCacheDto.class)
-			.map(dto -> Newsletter.create(dto.getId(), dto.getName(), dto.getDomain(),
-				dto.getMailingList(), dto.getStatus()));
+			.map(newsletterCacheDto -> newsletterCacheDto.toDomain().toEntityDto());
```
이부분에서 domain으로 바꿨다가 다시 entityDto로 바꾸는 이유는 뭘까요?

anaconda77 replied:
밑의 부분과 동일하게 EntityDTO로 응답하는 구조를 택하여서 그렇습니다..!

Gyu-won commented on src/main/java/com/newzet/api/mail/business/MailServiceFacade.java:
```diff
@@ -23,7 +23,7 @@ public class MailServiceFacade {
 		String htmlLink) {
 		User user = userService.getUserByEmail(toDomain);
-		Newsletter newsletter = newsletterService.findOrCreateNewsletter(fromName,
+		NewsletterEntityDto newsletter = newsletterService.findOrCreateNewsletter(fromName,
```
원래대로 Newsletter 도메인을 다루는게 좋아보이는데 EntityDto로 바꾸신 이유가 있나요?

anaconda77 replied:
지난 주에 Newsletter의 도메인 로직이 존재하지 않아서 굳이 Newsletter 도메인을 거쳐갈 필요가 있을까에 대한 의문이 들어 이 부분에 대해 논의할때, 도메인을 거치지 않고 EnityDTO에서 바로 Controller 응답 DTO로 변환해가는 구조도 가능하다는 얘기가 있었기에 해당 구조를 택하였습니다.
다만 제가 이해한건 도메인 로직이 없는 경우 굳이 도메인을 거치지 않아도 된다 였는데, 추후 도메인 로직이 추가될 잠재적인 상황을 고려한다면 도메인을 거치게 하는게 맞는 것인지 궁금합니다..!

Gyu-won commented on src/main/java/com/newzet/api/newsletter/business/NewsletterService.java:
```diff
@@ -34,18 +36,40 @@ public class NewsletterService {
 	private final LockFactory lockFactory;
 
-	public Newsletter findOrCreateNewsletter(String name, String domain, String mailingList) {
+	public NewsletterEntityDto findOrCreateNewsletter(String name, String domain, String mailingList) {
```
이부분도 Newsletter로 받는게 더 좋지 않을까요?

Gyu-won commented on src/main/java/com/newzet/api/newsletter/business/NewsletterService.java:
```diff
@@ -34,18 +36,40 @@ public class NewsletterService {
+	public NewsletterListResponse searchNewsletterListByNameOrCategoryId(String name,
+		String categoryId) {
+		UUID categoryUuid = UuidConverter.convert(categoryId);
+		List<NewsletterResponse> newsletterResponseList = newsletterRepository.findNewsLetterListByNameOrCategoryId(
+				name, categoryUuid).stream()
+			.map(newsletterEntity -> NewsletterResponse.create(newsletterEntity.getId(),
+				newsletterEntity.getName(), newsletterEntity.getImageUrl(),
+				newsletterEntity.getDescription(), newsletterEntity.getPriority()))
+			.toList();
```
이렇게 길어진다면 domain으로 변환 후 dto로 변환하는 것도 좋아보입니다! domain 변환을 하지 않는 이유가 복잡성을 줄이기 위함이었는데 이렇게 하니까 더 복잡해 보이네요! 다음에는 이런 경우 도메인으로 변환하는 것도 좋을 것 같아요!

Gyu-won commented on src/main/java/com/newzet/api/newsletter/business/NewsletterService.java:
```diff
@@ -34,18 +36,40 @@ public class NewsletterService {
 	}
 
-	private Newsletter findOrCreateByDomainOrMailingListWithLock(String name, String domain,
+	private NewsletterEntityDto findOrCreateByDomainOrMailingListWithLock(String name, String domain,
```
이부분도 앞에와 마찬가지로 Newsletter로 받으면 어떨까요?

Gyu-won commented on src/main/java/com/newzet/api/newsletter/repository/NewsletterEntity.java:
```diff
@@ -26,42 +34,96 @@
 
 	public NewsletterEntityDto toEntityDto() {
-		return NewsletterEntityDto.create(id, name, domain, mailingList, status.name());
+		return NewsletterEntityDto.create(id, name, category == null ? null : category.toEntityDto(), domain,
```
이 부분도 로직을 변경하면 꼭 수정해야겠네요..! 바꿀게 많군요!

anaconda77 replied:
네 현재 생성 로직에 생성 후 category get()시 null exception이 발생하는 문제가 있어서, 임시로 이렇게 조치해놓았습니다!

Gyu-won commented on src/main/java/com/newzet/api/newsletter/repository/NewsletterRepositoryImpl.java:
```diff
@@ -29,4 +31,28 @@ public Optional<NewsletterEntityDto> findByDomainOrMailingList(String domain,
+
+	//TODO: 다음 이슈에서 뉴스레터 추천 API 구현할때 사용할 예정
+	@Override
+	public List<NewsletterEntityDto> getNewsLetterListByCategoryIdList(List<UUID> categoryIdList) {
+		return newsletterJpaRepository.findAll().stream()
+			.filter(newsletterEntity -> categoryIdList.contains(newsletterEntity.getId()))
+			.map(NewsletterEntity::toEntityDto)
+			.toList();
```
이렇게 가져오는 것보다 DB에서 필터링해서 가져오는게 더 빠르지 않을까요?

anaconda77 replied:
categoryId 개수와 인덱싱을 고려하지 않았네요..DB에서 in절로 조회하는게 훨씬 빠른게 맞는 것 같습니다!

Gyu-won commented on src/main/java/com/newzet/api/subscription/business/service/SubscriptionRepository.java:
```diff
@@ -8,9 +8,9 @@
 
 public interface SubscriptionRepository {
-	SubscriptionEntityDto create(UserEntityDto userDto, NewsletterEntityDto newsletterDto);
+	SubscriptionEntityDto create(UserEntityDto userDto, NewsletterEntityDto newsletterEntity);
```
여기는 왜 NewsletterEntity로 변경되었나요?

anaconda77 replied:
네이밍을 실수한 것 같습니다 변경해야겠네요..😂

Gyu-won commented on src/main/java/com/newzet/api/subscription/business/service/SubscriptionService.java:
```diff
@@ -20,12 +20,12 @@ public class SubscriptionService {
 	private final SubscriptionRepository subscriptionRepository;
 
-	public void addSubscription(User user, Newsletter newsletter) {
+	public void addSubscription(User user, NewsletterEntityDto newsletter) {
```
이부분도 Newsletter Domain을 넘기는게 더 좋지 않을까요?

Gyu-won commented on src/main/java/com/newzet/api/subscription/repository/repository/SubscriptionRepositoryImpl.java:
```diff
@@ -22,23 +26,23 @@
-		NewsletterEntity newsletter = NewsletterEntity.create(newsletterDto.getId(),
-			newsletterDto.getName(), newsletterDto.getDomain(), newsletterDto.getMailingList(),
-			newsletterDto.getStatus());
-
+		UserEntity user = userJpaRepository.findById(userDto.getId())
+			.orElseThrow(() -> new NoUserException("해당 user를 찾을 수 없습니다."));
+		NewsletterEntity newsletter = newsletterJpaRepository.findById(newsletterDto.getId())
+			.orElseThrow(() -> new NoNewsletterException("해당 newsletter를 찾을 수 없습니다."));
```
이부분은 repository에서 조회하는 지금 방식이 더 좋은 것 같습니다! 이 방식에 맞게 저도 수정해야겠네요.

Gyu-won reviewed (approved):
바꾸실 부분이 엄청 많았겠네요..! 아직 Newsletter Insert 부분이 구현이 안돼서 많이 복잡하네요. 제가 Newsletter 수정 구현하면서 전체적으로 고치겠습니다. 정말 수고많으셨어요!

anaconda77 committed:
Refactor: getNewsletterListByCategoryIdList in절 쿼리로 동작하도록 변경

anaconda77 committed:
Refactor: Newsletter 서비스 단에서 EntityDto가 아닌 Domain을 받도록 수정
